File: gags.py
from math import ceil, floor
from collections import defaultdict
from pprint import pprint


class Cog:
    hp = (6, 12, 20, 30, 42, 56, 72, 90, 110, 132, 156, 200)

    def __init__(self, lvl):
        self.lvl = lvl

    def __repr__(self):
        return 'Cog(Lvl {} | {} Hp)'.format(self.lvl, self.max_hp())

    def max_hp(self):
        return self.hp[self.lvl-1]

    def to_JSON(self):
        return {
            'lvl': self.lvl,
            'hp': self.max_hp(),
        }


class Gag:
    gag_names = {
        'sound':    ('Bike Horn', 'Whistle', 'Bugle', 'Aoogah',
                     'Elephant Trunk', 'Foghorn', 'Opera Singer'),
        'throw':    ('Cupcake', 'Fruit Pie Slice', 'Cream Pie Slice', 'Whole Fruit Pie',
                     'Whole Cream Pie', 'Birthday Cake', 'Wedding Cake'),
        'squirt':   ('Squirting Flower', 'Glass of Water', 'Squirt Gun',
                     'Seltzer Bottle', 'Fire Hose', 'Storm Cloud', 'Geyser'),
        'drop':     ('Flower Pot', 'Sandbag', 'Anvil', 'Big Weight',
                     'Safe', 'Grand Piano', 'Toontanic'),
    }
    gag_dmgs = {
        'sound':    (4, 7, 11, 16, 21, 50, 90),
        'throw':    (6, 10, 17, 27, 40, 100, 120),
        'squirt':   (4, 8, 12, 21, 30, 80, 105),
        'drop':     (10, 18, 30, 45, 60, 170, 180),
    }

    def __init__(self, track, lvl, org=False):
        self.track = track
        self.lvl = lvl
        self.org = org

    def __repr__(self):
        if self.lvl == 0:
            return '({}, {}'.format('None', 0)
        return '({}{}, {})'.format('[org] ' if self.org else '', self.name(), self.base_dmg())

    def __eq__(self, other):
        return (self.lvl == other.lvl and self.org == other.org and
                self.track == other.track)

    def to_JSON(self):
        return {
            'name':  self.name(),
            'lvl':   self.lvl,
            'org':   self.org,
            'track': self.track,
        }

    def name(self):
        if self.lvl == 0:
            return 'None'
        return self.gag_names[self.track][self.lvl-1]

    def base_dmg(self):
        if self.lvl == 0:
            return 0
        # If gag dmg is less than 10, multiplying by 1.1 does not result in
        # the minimum 1 bonus dmg point. Ex: 9*1.1=9.9 floor(9.9) == 9
        if self.org and 0 < self.gag_dmgs[self.track][self.lvl-1] < 10:
            return self.gag_dmgs[self.track][self.lvl-1] + 1
        # Normal organic calculation, 10% dmg bonus
        return (floor(self.gag_dmgs[self.track][self.lvl-1] *
                      (1.1 if self.org else 1)))

    # When doing track count check for knockback multiplier, if both
    # 'throw' and 'squirt' are present, only grant the 0.5 knockback
    # multiplier to 'throw'. Currently not an issue.
    def multiplier(self, combo):
        if self.lvl == 0:
            return 1  # Could be 0 also... zero times zero equals...
        tc = combo.track_counts()
        multi = (1 + (0.2 if tc[self.track] >= 2 else 0) +
                 (0.5 if combo.lured and self.track in {'throw', 'squirt'} else 0))
        return multi


class Combo:
    def __init__(self, gags, num, lured, sel_track, sel_stun):
        self.gags = gags  # list of gags
        self.num = num  # number of gags
        self.lured = lured  # lured - True or False
        self.sel_track = sel_track
        self.sel_stun = sel_stun   # '' unless sel_track == 'drop'

    def __repr__(self):
        return 'Combo({})'.format(self.gags)

    # Returns a dict to be JSON serializable, containing combo info
    def to_JSON(self):
        json_ = {
            'num': self.num,
            'sel_track': self.sel_track,
            'sel_stun': self.sel_stun,
            'lured': self.lured,
            'damage': self.damage(),
            'gags': [gag.to_JSON() for gag in self.gags],
        }
        return json_

    def track_counts(self):
        tc = defaultdict(int)
        for gag in self.gags:
            if gag.lvl != 0:
                tc[gag.track] += 1
        return tc

    def damage(self):
        dmg = 0
        for gag in self.gags:
            dmg += round(gag.base_dmg() * gag.multiplier(self), 2)
        return ceil(dmg)


class GagCombos:
    def __init__(self, cog_lvl, sel_track, num, lured=False, sel_stun=''):
        self.cog = Cog(cog_lvl)
        self.sel_track = sel_track
        self.num = num
        self.lured = lured
        self.sel_stun = sel_stun if self.sel_track == 'drop' else ''
        # Avoids bug of sel_stun accidentally being set when track != 'drop'.

    def find(self):
        combo = self.test_down(self.test_up())
        return combo

    def to_JSON(self):
        json_ = self.find().to_JSON()  # find() returns a combo, which is then converted to JSON
        json_['cog_lvl'] = self.cog.lvl
        return json_

    def test_up(self):
        # Begin with lvl 1 gags
        if self.sel_track == 'drop':
            combo = Combo([Gag(self.sel_stun, 1)] +
                          [Gag(self.sel_track, 1) for i in range(self.num-1)],
                          self.num, self.lured, self.sel_track, self.sel_stun)
        else:
            combo = Combo([Gag(self.sel_track, 1) for i in range(self.num)],
                          self.num, self.lured, self.sel_track, self.sel_stun)
        while combo.damage() < self.cog.max_hp():
            for gag in combo.gags:
                if gag.track == self.sel_stun and gag.lvl > 4:
                    gag.lvl = 5
                else:
                    gag.lvl += 1
        return combo

    def test_down(self, combo):
        i = 0  # Arbitrary value
        # While the last gag in the list was not the one last modified
        while i != len(combo.gags) - 1:
            # Begins with index at end of gag list, decrements to reach start index 0
            for i in range(len(combo.gags) - 1, -1, -1):
                if combo.gags[i].lvl == 0:  # A lvl 0 gag is None
                    break
                combo.gags[i].lvl -= 1  # lower gag lvl
                if combo.damage() < self.cog.max_hp():  # test if new dmg is sufficient
                    combo.gags[i].lvl += 1  # revert gag lvl change
                    break
        # Attempt to lower gag lvl of stun gag in a drop combo
        if self.sel_track == 'drop' and combo.gags[0].lvl == 5:
            if combo.gags[1].lvl == 7:  # if lvl 7 gag is being used, try alternative
                combo.gags[0].lvl += 1  # set stun to lvl 6 gag from 5
                combo.gags[1].lvl -= 1  # set gag to lvl 6 from 7
                # If new combo dmg is insufficient, revert
                if combo.damage() < self.cog.max_hp():
                    combo.gags[0].lvl -= 1
                    combo.gags[1].lvl += 1
            else:
                # Lower lvl of stun gag while damage is sufficient or gag becomes None
                while combo.damage() >= self.cog.max_hp() and combo.gags[0].lvl > 0:
                    combo.gags[0].lvl -= 1
                combo.gags[0].lvl += 1
        return combo


if __name__ == '__main__':
    # Test as you wish
    combo = GagCombos(cog_lvl=11, sel_track='drop', num=3, lured=True, sel_stun='squirt').find()
    print(combo, "\n")  # Combo __repr__
    combo_json = combo.to_JSON()  # Combo as JSON
    print("Cog HP: {}".format(Cog(11).max_hp()))
    pprint(combo_json)

File: gags.js
export const classicCogHp = {
  1:  6,
  2:  12,
  3:  20,
  4:  30,
  5:  42,
  6:  56,
  7:  72,
  8:  90,
  9:  110,
  10: 132,
  11: 156,
  12: 200,
};

export const ttrCogHp = {
  ...classicCogHp,
  12: 196,
  13: 224,
  14: 254,
  15: 286,
  16: 320,
  17: 356,
  18: 394,
  19: 434,
  20: 476,
};

const gagTracks = {
  'toonup': 'toonup',
  'trap': 'trap',
  'lure': 'lure',
  'sound': 'sound',
  'throw': 'throw',
  'squirt': 'squirt',
  'drop': 'drop',
};

const gagTracksOrder = {
  [gagTracks.toonup]: 0,
  [gagTracks.trap]: 1,
  [gagTracks.lure]: 2,
  [gagTracks.sound]: 3,
  [gagTracks.throw]: 4,
  [gagTracks.squirt]: 5,
  [gagTracks.drop]: 6,
};

const gags = {
  'trap': {
    1: { name: 'Banana Peel',  damage: 12  },
    2: { name: 'Rake',         damage: 20  },
    3: { name: 'Marbles',      damage: 35  },
    4: { name: 'Quicksand',    damage: 50  },
    5: { name: 'Trapdoor',     damage: 85  },
    6: { name: 'TNT',          damage: 180 },
    7: { name: 'Railroad',     damage: 200 },
  },
  'sound': {
    1: { name: 'Bike Horn',       damage: 4  },
    2: { name: 'Whistle',         damage: 7  },
    3: { name: 'Bugle',           damage: 11 },
    4: { name: 'Aoogah',          damage: 16 },
    5: { name: 'Elephant Trunk',  damage: 21 },
    6: { name: 'Foghorn',         damage: 50 },
    7: { name: 'Opera Singer',    damage: 90 },
  },
  'throw': {
    1: { name: 'Cupcake',          damage: 6   },
    2: { name: 'Fruit Pie Slice',  damage: 10  },
    3: { name: 'Cream Pie Slice',  damage: 17  },
    4: { name: 'Whole Fruit Pie',  damage: 27  },
    5: { name: 'Whole Cream Pie',  damage: 40  },
    6: { name: 'Birthday Cake',    damage: 100 },
    7: { name: 'Wedding Cake',     damage: 120 },
  },
  'squirt': {
    1: { name: 'Squirting Flower',  damage: 4   },
    2: { name: 'Glass of Water',    damage: 8   },
    3: { name: 'Squirt Gun',        damage: 12  },
    4: { name: 'Seltzer Bottle',    damage: 21  },
    5: { name: 'Fire Hose',         damage: 30  },
    6: { name: 'Storm Cloud',       damage: 80  },
    7: { name: 'Geyser',            damage: 105 },
  },
  'drop': {
    1: { name: 'Flower Pot',   damage: 10  },
    2: { name: 'Sandbag',      damage: 18  },
    3: { name: 'Anvil',        damage: 30  },
    4: { name: 'Big Weight',   damage: 45  },
    5: { name: 'Safe',         damage: 60  },
    6: { name: 'Grand Piano',  damage: 170 },
    7: { name: 'Toontanic',    damage: 180 },
  },
};

const ttrGagsDmgOverrides = {
  'drop': {
    5: { name: 'Safe', damage: 70 },
  },
};

export class Cog {
  /**
   * @param {Object} args
   * @param {number} args.lvl
   * @param {boolean} args.isLured
   * @param {boolean} [args.isV2=false]
   * @param {string} args.game
   */
  constructor({ lvl, isLured, game, isV2 = false }) {
    this.lvl = lvl;
    this.isLured = isLured;
    this.isV2 = isV2;
    this._game = game;
  }

  /**
   * @return {number}
   */
  get hp() {
    return (this._game === 'ttr' ? ttrCogHp : classicCogHp)[this.lvl];
  }

  /**
   * @return {boolean}
   */
  get hasReinforcedPlating() {
    return this.isV2 && this._game === 'ttr';
  }

  /**
   * Get the amount of damage resisted per gag,
   * if the cog has reinfoced plating, else zero.
   */
  get reinforcedPlatingDmgResist() {
    if (!this.hasReinforcedPlating) return 0;
    return Math.floor(this.lvl * 1.5)
  }

  /**
   * @return {string}
   */
  toString() {
    return `Cog(level: ${this.lvl}, HP: ${this.hp})`;
  }
}

export class Gag {
  // TODO kwargs with obj
  /**
   * @param {string} track
   * @param {number} lvl
   * @param {string} game
   * @param {boolean} [isOrg=false]
   */
  constructor(track, lvl, game, isOrg = false) {
    this.track = track;
    this.lvl = lvl;
    this._game = game;
    this.isOrg = isOrg;
  }

  /**
   * @return {string}
   */
  get name() {
    if (this.lvl === 0) {
      return 'None';
    }
    return gags[this.track][this.lvl].name;
  }

  /**
   * @return {number}
   */
  get damage() {
    if (this.lvl === 0) {
      return 0;
    }
    const baseDmg = this._game === 'ttr'
      ? (ttrGagsDmgOverrides[this.track]?.[this.lvl]?.damage ?? gags[this.track][this.lvl].damage)
      : gags[this.track][this.lvl].damage;

    if (this.isOrg) {
      if (baseDmg < 10) return baseDmg + 1;
      else return Math.floor(baseDmg * 1.1);
    } else {
      return baseDmg;
    }
  }

  /**
   * @param {Combo} combo
   * @param {boolean} isCogLured
   * @return {number}
   */
  multiplier(combo, isCogLured) {
    if (this.lvl === 0)
      return 1;
    // No multipliers applicable for trap.
    // But multiple trap should not be present anyway in the combo.
    if (this.track === 'trap')
      return 1;
    const tc = combo.trackCounts();
    return (
      1
      + (tc[this.track] >= 2 ? 0.2 : 0)
      + (this.knockback(combo, isCogLured) ? 0.5 : 0)
    );
  }

  /**
   * @param {Combo} combo
   * @param {boolean} isCogLured
   * @return {boolean}
   */
  knockback(combo, isCogLured) {
    if (!isCogLured)
      return false;
    const tc = combo.trackCounts();
    if (tc['trap'] > 0)
      return false;
    if (tc['sound'] > 0)
      return false;
    if (this.track === 'throw')
      return true;
    if (this.track === 'squirt' && tc['throw'] === 0)
      return true;
    return false;
  }

  /**
   * @return {string}
   */
  toString() {
    return `Gag(name: ${this.name}, level: ${this.lvl}, damage: ${this.damage}, isOrg: ${this.isOrg})`;
  }
}

export class Combo {
  /**
   * @param {Object} args
   * @param {Array<Gag>} args.gags
   */
  constructor({
    gags,
  }) {
    this.gags = gags;
  }

  /**
   * @return {Object<string, number>} A mapping of gag tracks to their respective number of gags in the combo
   */
  trackCounts() {
    return this.gags.reduce((acc, gag) => {
      if (gag.lvl !== 0) {
        acc[gag.track] += 1;
      }
      return acc;
    }, {
      'toonup': 0,
      'trap': 0,
      'lure': 0,
      'sound': 0,
      'throw': 0,
      'squirt': 0,
      'drop': 0,
    });
  }

  /**
   * @param {Cog} cog
   * @return {number}
   */
  damage(cog) {
    const dmg = this.gags.reduce((dmg, gag) => {
      return dmg + (
        (gag.damage - cog.reinforcedPlatingDmgResist)
        * gag.multiplier(this, cog.isLured)
      );
    }, 0);
    return Math.ceil(dmg);
  }

  /**
   * @param {Cog} cog
   * @return {boolean}
   */
  damageKillsCog(cog) {
    return this.damage(cog) >= cog.hp;
  }

  /**
   * @return {string}
   */
  toString() { // TODO
    return `Combo(damage: ${this.damage()} gags: [\n${this.gags.map(g => `  ${g}`).join(',\n')}\n])`;
  }
}

/**
 * @param {Object} args
 * @param {number} args.cogLvl
 * @param {boolean} args.isLured
 * @param {Object<string, number>} args.gags
 *   Mapping of gag track to number of gags to use in the combo. Sum of numbers must be 1 <= n <= 4.
 * @param {Object<string, number>} args.organicGags
 * @param {string} args.game
 * @return {Array<Combo>}
 */
export function findCombo({
  cogLvl,
  isLured,
  gags,
  organicGags,
  game = 'ttr',
}) {
  let sumGags = 0;
  for (const [key, val] of Object.entries(gags)) {
    if (!gagTracks.hasOwnProperty(key)) {
      throw new Error(
        `Unrecognized gag track key '${key}' in \`gags\` argument. ` +
        `Must be one of: ${Object.keys(gagTracks).map(gt => "'" + gt + "'").join(', ')}`
      );
    }
    if (typeof val !== 'number' || isNaN(val)) {
      throw new Error(`Values in \`gags\` argument must be a number. For key '${key}' got value '${val}'.`);
    }
    if (key === 'trap' && val > 1) {
      throw new Error(`Cannot specify multiple trap gags, must be 0 or 1.`);
    }
    sumGags += val;
  }
  if (sumGags < 1 || sumGags > 4) {
    throw new Error(`Sum of values in \`gags\` argument must be 1 <= n <= 4. Received ${sumGags}.`);
  }

  /** @type {Array<Gag>} */
  const comboGags = Object.entries(gags).reduce((acc, [gagTrack, numGags]) => {
    for (let i = 0; i < numGags; i++) {
      const isOrg = i <= (organicGags[gagTrack] ?? 0) - 1;
      acc.push(new Gag(gagTrack, 1, game, isOrg));
    }
    return acc;
  }, []);

  // Sort by damage high to low for the algorithm
  // comboGags.sort((gag1, gag2) => gag2.damage - gag1.damage);
  comboGags.sort((gag1, gag2) => gag1.damage - gag2.damage);

  const combo = new Combo({
    game,
    gags: comboGags,
    organicGags,
  });
  // Increase the level of each gag until the damage is sufficient
  while (!combo.damageKillsCog(cog)) {
    for (const gag of combo.gags) {
      if (gag.lvl === 7) {
        continue;
      } else {
        gag.lvl += 1;
      }
    }
    if (combo.gags.every(gag => gag.lvl === 7))
      break;
  }

  // Insufficient damage, killing cog with given parameters is not possible
  if (!combo.damageKillsCog(cog)) {
    combo.gags = combo.gags.map(g => new Gag(g.track, 0, game));
    return combo;
  }

  // Decrease the level of individual gags until the damage is insufficient
  // TODO Maybe combine the org part here? Nah above
  let i = 0;
  while (i != combo.gags.length-1) {
    for (i = combo.gags.length-1; i >= 0; i--) {
      if (combo.gags[i].lvl === 0) {
        break;
      }
      combo.gags[i].lvl -= 1;
      if (!combo.damageKillsCog(cog)) {
        combo.gags[i].lvl += 1;
        break
      }
    }
  }

  // Check if organic is actually necessary for the combo
  // if (gagTrack === 'drop' && isStunOrg) {
  //   combo.gags[0].isOrg = false;
  //   if (!combo.damageKillsCog())
  //     combo.gags[0].isOrg = true;
  // }
  // TODO refactor
  for (let i = 0; i < combo.gags.length; i++) {
    combo.gags[i].isOrg = false;
    if (!combo.damageKillsCog()) {
      combo.gags[i].isOrg = true;
    }
  }

  // TODO The approach to generating multiple combos
  // might be to just handle level 6 & 7 gags differently.
  // For example, if a combo for '1 sound, 1 throw, 1 squirt'
  // contains a fog, try regenerating with no fog, i.e with a
  // level 6 throw in one combo and a level 6 squirt in another!

  combo.gags.sort(sortFnGags);

  const combos = [combo];

  return combos;
}

/**
 * Sort gags based on track order, then damage high to low.
 *
 * @param {Gag} gag1
 * @param {Gag} gag2
 * @return {number}
 */
function sortFnGags(gag1, gag2) {
  const cmpTrack = gagTracksOrder[gag1.track] - gagTracksOrder[gag2.track];
  if (cmpTrack !== 0) return cmpTrack;
  return gag2.damage - gag1.damage;
}

/**
 * @param {Combo} combo
 */
export function logTable(combo) {
  console.log('Input:');
  console.table({
    'Game': combo.game === 'ttr' ? 'Toontown Rewritten' : 'Toontown Online (Classic)',
    'Num Toons': combo.numToons,
    'Cog Level': combo.cog.lvl,
    'Is Cog Lured?': combo.isLured,
    'Gag Track': combo.gagTrack,
    ...(combo.gagTrack === 'drop' && { 'Stun Gag Track': combo.stunTrack }),
    'Organic Gags': Object.entries(combo.organicGags)
      .map(([track, num]) => !num ? null : `${track[0].toUpperCase()}${track.slice(1)}: ${num}`)
      .filter(s => s !== null)
      .join(', '),
  });
  console.log('Combo:');
  console.table(
    combo.gags.reduce((newObj, gag, i) => {
      newObj[`Gag ${i+1}`] = {
        'Name': `${gag.name}${gag.isOrg ? ' (Organic)' : ''}`,
        'Level': gag.lvl,
        'Damage': gag.damage,
      };
      return newObj;
    }, {})
  );
  // TODO Maybe have a `multipliers()` method on `Combo` that returns this,
  // providing additional info on which track(s) have multipliers added to them.
  const multipliers = [];
  if (combo.trackCounts()[combo.gagTrack] >= 2)
    multipliers.push('Same type bonus (20%)'); // TODO Nope, not for trap. Delegate to Combo class
  if (combo.isLured && [combo.gagTrack, combo.stunTrack].some(track => new Set(['throw', 'squirt']).has(track)))
    multipliers.push('Knockback bonus (50%)');
  console.log('Details:');
  console.table({
    'Total base damage': combo.gags.reduce((sum, gag) => sum + gag.damage, 0),
    'Damage Multipliers': multipliers.join(', ') || '<None>',
    'Final damage calculated': combo.damage(), // TODO
    'Cog HP': combo.cog.hp,
  });
}

export default findCombo;
